packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -440,11 +441,105 @@ class RenderParagraph extends RenderBox {
     super.describeSemanticsConfiguration(config);
-    config
-      ..label = text.toPlainText()
-      ..textDirection = textDirection;
+    bool hasRecognizers = false;
+    text.visitTextSpan((TextSpan span) {
+      if (span.recognizer != null) {
+        hasRecognizers = true;
+        return false;
+      }
+      return true;
+    });
+    if (hasRecognizers) {
+      config.explicitChildNodes = true;
+      config.isSemanticBoundary = true;
+    } else {
+      config.label = text.toPlainText();
+      config.textDirection = textDirection;
+    }
   }
 
+  @override
+  void assembleSemanticsNode(SemanticsNode node, SemanticsConfiguration config, Iterable<SemanticsNode> children) {
+    final List<SemanticsNode> newChildren = <SemanticsNode>[];
+    StringBuffer buffer;
+    double order = 0.0;
+    int start = 0;
+    int end = 0;
+    // traverse all spans and collect them into adjacent compatible nodes.
+    text.visitTextSpan((TextSpan span) {
+      if (span.recognizer != null) {
+        if (buffer != null) {
+          final SemanticsNode child = new SemanticsNode();
+          final SemanticsConfiguration childConfig = new SemanticsConfiguration();
+          child.rect = _getRectForSpan(start, end);
+          childConfig
+            ..label = buffer.toString()
+            ..textDirection = textDirection
+            ..sortKey = new OrdinalSortKey(order++);
+          buffer = null;
+          child.updateWith(config: childConfig);
+          newChildren.add(child);
+          start = end;
+        }
+        end += span.text.length;
+        final SemanticsNode child = new SemanticsNode();
+        final SemanticsConfiguration childConfig = new SemanticsConfiguration();
+        childConfig.textDirection = textDirection;
+        childConfig.label = span.text;
+        childConfig.sortKey = new OrdinalSortKey(order++);
+        if (span.recognizer is LongPressGestureRecognizer) {
+          final LongPressGestureRecognizer recognizer = span.recognizer;
+          childConfig.onLongPress = recognizer.onLongPress;
+        } else if (span.recognizer is TapGestureRecognizer) {
+          final TapGestureRecognizer recognizer = span.recognizer;
+          childConfig.onTap = recognizer.onTap;
+        }
+        child.rect = _getRectForSpan(start, end);
+        child.updateWith(config: childConfig);
+        newChildren.add(child);
+        start = end;
+      } else {
+        buffer ??= new StringBuffer();
+        buffer.write(span.text);
+        end += span.text.length;
+      }
+      return true;
+    });
+    if (buffer != null && buffer.isNotEmpty) {
+      final SemanticsNode child = new SemanticsNode();
+      final SemanticsConfiguration childConfig = new SemanticsConfiguration();
+      child.rect = _getRectForSpan(start, end);
+      childConfig
+        ..sortKey = new OrdinalSortKey(order++)
+        ..label = buffer.toString()
+        ..textDirection = textDirection;
```

Comment:
i think this code appears more or less verbatim three times, which suggests we should factor it out

Reply:
Done